File: matlab/+Llt/+DemoBoardExamples/+LTC2000/Ltc2000Dc2085.m
% DC2085 / LTC2000 Interface Example
% LTC2000: 16-/14-/11- Bit 2.5 Gsps DAcs
%
% This program demonstrates how to communicate with the LTC2000 demo board 
% using Matlab. Examples are provided for generating sinusoidal data from within
% the program, as well as writing and reading pattern data from a file.
% 
% Board setup is described in Demo Manual 2085. Follow the procedure in 
% this manual, and verify operation with the LTDACGen software. Once 
% operation is verified, exit LTDACGen and run this script.
% 
% Demo board documentation:
% http://www.linear.com/demo/2085
% http://www.linear.com/product/LTC2000#demoboards
% 
% LTC2000 product page
% http://www.linear.com/product/LTC2000
%  
% REVISION HISTORY
% $Revision$
% $Date$
%
% Copyright (c) 2015, Linear Technology Corp.(LTC)
% All rights reserved.
% 
% Redistribution and use in source and binary forms, with or without
% modification, are permitted provided that the following conditions are met:
% 
% 1. Redistributions of source code must retain the above copyright notice, 
%    this list of conditions and the following disclaimer.
% 2. Redistributions in binary form must reproduce the above copyright notice,
%    this list of conditions and the following disclaimer in the documentation
%    and/or other materials provided with the distribution.
% 
% THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
% ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
% WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
% DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
% ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
% (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
% LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
% ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
% (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
% SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
% 
% The views and conclusions contained in the software and documentation are those
% of the authors and should not be interpreted as representing official policies,
% either expressed or implied, of Linear Technology Corp.

% NOTE:
% 	ADD THE ABSOLUTE PATH TO "linear_lab_tools\matlab" FOLDER BEFORE RUNNING THE SCRIPT.
%   RUN "mex -setup" TO SET UP COMPILER AND CHOSE THE OPTION "Lcc-win32 C".

function Ltc2000Dc2085(arg1NumSamples, arg2Verbose)

    AMPLITUDE = 16000;
    NUM_CYCLES = 800;   % Number of sinewave cycles over the entire data record
    SLEEP_TIME = 0.1;

    if(~nargin)
        numSamples = 65536;    % n.BuffSize
        % Print extra information to console
        verbose = true;
    else
        numSamples = arg1NumSamples;
        verbose = arg2Verbose;
    end

    if verbose
        fprintf('LTC2000 Test Script\n');
    end

    % Import LTC2000 definitions and support functions
    lt2k = Llt.DemoBoardExamples.LTC2000.Ltc2000Constants();

    % Returns the object in the class constructor
    lths = Llt.Common.LtcControllerComm();

    deviceInfoList = lths.ListControllers(lths.TYPE_HIGH_SPEED);
    deviceInfo = [];
    for info = deviceInfoList
       if strcmp(info.description(1:7), 'LTC2000')
           deviceInfo = info;
       end
    end

    if isempty(deviceInfo)
        error('TestLtcHighSpeedComm:noDevice', 'No LTC200 demo board detected');
    end

    fprintf('Found LTC2000 demo board:\n');
    fprintf('Description: %s\n', deviceInfo.description);
    fprintf('Serial Number: %s\n', deviceInfo.serialNumber);

    % init a device and get an id
    did = lths.Init(deviceInfo);

    lths.HsSetBitMode(did, lths.HS_BIT_MODE_MPSSE);
    lths.HsFpgaToggleReset(did);

    fprintf('FPGA Load ID is %X\n', lths.HsFpgaReadDataAtAddress(did, lt2k.FPGA_ID_REG));
    fprintf('Reading PLL status, should be 0x47\n');
    data = lths.HsFpgaReadDataAtAddress(did, lt2k.FPGA_STATUS_REG);
    fprintf('And it is... 0x%s\n', dec2hex(data));
    fprintf('Turning on DAC...\n');
    
    lths.HsFpgaWriteDataAtAddress(did, lt2k.FPGA_DAC_PD, 1);

    pause(SLEEP_TIME);

    if verbose
        fprintf('Configuring ADC over SPI\n');
    end

    % Initial register values can be taken directly from LTDACGen.
    % Refer to LTC2000 datasheet for detailed register descriptions.

    SpiWrite(lt2k.REG_RESET_PD, 0);
    SpiWrite(lt2k.REG_CLK_CONFIG, 2);    % setting output current to 7mA?
    SpiWrite(lt2k.REG_CLK_PHASE, 7);     % DCKIP/N delay = 315 ps
    SpiWrite(lt2k.REG_PORT_EN, 11);      % Enables Port A and B Rxs and allows DAC to be updated from A and B
    SpiWrite(lt2k.REG_SYNC_PHASE, 0);
    SpiWrite(lt2k.REG_LINER_GAIN, 0);
    SpiWrite(lt2k.REG_LINEARIZATION, 8);
    SpiWrite(lt2k.REG_DAC_GAIN, 32);
    SpiWrite(lt2k.REG_LVDS_MUX, 0);
    SpiWrite(lt2k.REG_TEMP_SELECT, 0);
    SpiWrite(lt2k.REG_PATTERN_ENABLE, 0);

    pause(SLEEP_TIME);

    % Optionally read back all registers
    if verbose
        fprintf('LTC2000 Register Dump:\n');
        for k = 0:9
            fprintf('Register %d: 0x%02X\n', k, SpiRead(k));
        end
    end

    lths.HsFpgaWriteDataAtAddress(did, lt2k.FPGA_CONTROL_REG, 32);

    pause(SLEEP_TIME);

    % Demonstrates how to generate sinusoidal data. Note that the total data 
    % record length contains an exact integer number of cycles.
    data = round(AMPLITUDE * sin((NUM_CYCLES * 2 * pi / numSamples) * ...
        (0:(numSamples - 1))));

    % Demonstrates how to generate sinc data.
    sincData = zeros(numSamples, 1);
    for i = 1:numSamples
    x = ((i - 32768) / (512.0)) + 0.0001;
    sincData(i) = int16((32000 * (sin(x) / x)));
    end

    % Demonstrates how to write generated data to a file.
    fprintf('writing data out to file')
    outFile = fopen('dacdata_sinc.csv', 'w');
    for i = 1 : numSamples
        fprintf(outFile, '%d\n', sincData(i));
    end
    fclose(outFile);
    fprintf('\ndone writing!')

    % Demonstrates how to read data in from a file.
    inData = zeros(numSamples, 1);
    fprintf('\nreading data from file')
    inFile = fopen('dacdata_sinc.csv', 'r');
    for i = 1: numSamples 
        inData(i) = str2double(fgetl(inFile));
    end
    fclose(inFile);
    fprintf('\ndone reading!')

    lths.HsSetBitMode(did, lths.HS_BIT_MODE_FIFO);
    % DAC should start running here!
    numBytesSent = lths.DataSendUint16Values(did, inData);
    fprintf('\nnumBytesSent (should be %d) = %d\n', numSamples * 2, ...
        numBytesSent);
    fprintf('You should see a waveform at the output of the LTC2000 now!\n');
    
    %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
    %%%    FUNCTION DEFINITIONS    
    %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
    
    function SpiWrite(address, value)
        lths.SpiSendByteAtAddress(did, bitor(address, lt2k.SPI_WRITE), value);
    end

    function value = SpiRead(address)
        value = lths.SpiReceiveByteAtAddress(did, bitor(address, lt2k.SPI_READ));
    end

end

File: python/llt/demo_board_examples/ltc2000/ltc2000_dc2085.py
# -*- coding: utf-8 -*-
"""
    E-mail: quikeval@linear.com

    Copyright (c) ?year?, Linear Technology Corp.(LTC)
    All rights reserved.

    Redistribution and use in source and binary forms, with or without
    modification, are permitted provided that the following conditions are met:

    1. Redistributions of source code must retain the above copyright notice, 
       this list of conditions and the following disclaimer.
    2. Redistributions in binary form must reproduce the above copyright 
       notice, this list of conditions and the following disclaimer in the 
       documentation and/or other materials provided with the distribution.

    THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
    AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE 
    IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE 
    ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE 
    LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
    CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF 
    SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS 
    INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN 
    CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
    ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE 
    POSSIBILITY OF SUCH DAMAGE.

    The views and conclusions contained in the software and documentation are 
    those of the authors and should not be interpreted as representing official
    policies, either expressed or implied, of Linear Technology Corp.

    Description:
        The purpose of this module is to demonstrate how to communicate with 
        the LTC2000 demo board through python with the Altera Stratix IV GX 
        FPGA Development Kit.
"""
import llt.common.ltc_controller_comm as comm
import llt.common.constants as consts
import llt.common.functions as funcs
import llt.common.exceptions as errs
import math
import time

def ltc2000_dc2085(data, spi_regs, verbose=False):
    with Ltc2000(False, spi_regs, verbose) as controller:
        controller.send_data(data)

class Ltc2000():
    _SIZE_DICTIONARY = {16*1024: 0x00, 32*1024: 0x10, 64*1024: 0x20, 128*1024: 0x30,
                        256*1024: 0x40, 512*1024: 0x50, 1024*1024: 0x60, 2*1024*1024: 0x70,
                        4*1024*1024: 0x80, 8*1024*1024: 0x90, 16*1024*1024: 0xA0,
                        32*1024*1024: 0xB0, 64*1024*1024: 0xC0, 128*1024*1024: 0xD0,
                        256*1024*1024: 0xE0, 512*1024*1024: 0xF0}
    _FPGA_ID_REG = 0x00
    _FPGA_CONTROL_REG = 0x01
    _FPGA_STATUS_REG = 0x02
    _FPGA_DAC_PD = 0x03
    
    def __init__(self, is_xilinx, spi_reg_values, verbose = False):
        if is_xilinx:
            self.expected_description = "LTC Communication Interface"
            self.expected_id = 0x20
            self.expected_pll_status = 0x06
            self.expected_max_val = 2 * 1024 * 1024
            self.range_string = "16K to 2M"
            self.is_little_endian = True;
        else:
            self.expected_description = "LTC2000"
            self.expected_id = 0x1A
            self.expected_pll_status = 0x47
            self.expected_max_val = 512 * 1024 * 1024
            self.range_string = "16K to 512M"
            self.is_little_endian = False;
        self.vprint = funcs.make_vprint(verbose)
        self._connect()
        self._init_controller(spi_reg_values)

    # support "with" semantics
    def __enter__(self):
        return self

    # support "with" semantics
    def __exit__(self, vtype, value, traceback):
        del vtype
        del value
        del traceback
        self.controller.close()
        self.controller.cleanup()

    def _connect(self):
        # Open communication to the demo board
        self.vprint("Looking for Controller")
        for info in comm.list_controllers(consts.TYPE_HIGH_SPEED):
            description = info.get_description()
            if self.expected_description in description:
                self.vprint("Found a possible setup")
                self.controller = comm.Controller(info)
                return
        raise(errs.HardwareError('Could not find a compatible device'))
    
    def _init_controller(self, spi_reg_values):
        self.controller.hs_set_bit_mode(consts.HS_BIT_MODE_MPSSE)
        self.controller.hs_fpga_toggle_reset()
        # Read FPGA ID register
        id = self.controller.hs_fpga_read_data_at_address(Ltc2000._FPGA_ID_REG)
        self.vprint("FPGA Load ID: 0x{:04X}".format(id))
        if self.expected_id != id:
            raise(errs.HardwareError('Wrong FPGA Load'))
        self.controller.hs_fpga_write_data_at_address(Ltc2000._FPGA_DAC_PD, 0x01)
        self.set_spi_registers(spi_reg_values)

    def set_spi_registers(self, register_values):
        self.vprint("Configuring ADC over SPI")
        if register_values != []:
            for x in range(0,len(register_values), 2):
                self.controller.spi_send_byte_at_address(register_values[x], register_values[x+1])

    def send_data(self, data):
        num_samples = len(data)                  
        num_samples_reg_value = Ltc2000._SIZE_DICTIONARY.get(num_samples)
        if num_samples > self.expected_max_val or num_samples_reg_value is None:
            raise(errs.NotSupportedError(
                "Data Length Not Supported (Must be a power of 2 between " + self.range_string))
        
        self.vprint("Reading PLL status")
        pll_status = self.controller.hs_fpga_read_data_at_address(Ltc2000._FPGA_STATUS_REG)        
        if self.expected_pll_status != pll_status:
            raise(errs.HardwareError('FPGA PLL status was bad'))
        self.vprint("PLL status is okay")
        time.sleep(0.1)
        self.controller.hs_fpga_write_data_at_address(Ltc2000._FPGA_CONTROL_REG, num_samples_reg_value)
        if self.is_little_endian:
            self.controller.data_set_low_byte_first()
        else:
            self.controller.data_set_high_byte_first()
        
        self.controller.hs_set_bit_mode(consts.HS_BIT_MODE_FIFO)
        self.vprint("Sending data")
        num_bytes_sent = self.controller.data_send_uint16_values(data)
        self.controller.hs_set_bit_mode(consts.HS_BIT_MODE_MPSSE)
        if num_bytes_sent != num_samples * 2:
            raise(errs.HardwareError("Not all data was sent."))
        self.vprint("All data was sent (" + str(num_bytes_sent) + " bytes)")

if __name__ == '__main__':
    num_cycles = 800  # Number of sine wave cycles over the entire data record
    total_samples = 64 * 1024 
    data = total_samples * [0] 

    for i in range(0, total_samples):
        data[i] = int(32000 * math.sin((num_cycles*2*math.pi*i)/total_samples))

    spi_regs = [ # addr, value
                   0x01, 0x00, 
                   0x02, 0x02, 
                   0x03, 0x07, 
                   0x04, 0x0B, 
                   0x05, 0x00, 
                   0x07, 0x00,
                   0x08, 0x08, 
                   0x09, 0x20, 
                   0x18, 0x00, 
                   0x19, 0x00,
                   0x1E, 0x00
               ]
    # to use this function in your own code you would typically do
    # ltc2000_dc2085(data, spi_reg)
    ltc2000_dc2085(data, spi_regs, verbose=True)
